### ISSUE TITLE
CMC crashing at first point in a trajectory

### ISSUE THREAD

**Issue by SietseAchterop** [opensim-org/opensim-core#3183]
Hello all,
been struggling with this for quite some time. The CMC-tool  crashes immediately with this model.

`EXIT: Converged to a point of local infeasibility. Problem may be infeasible.
[error] SimTK Exception thrown at InteriorPointOptimizer.cpp:264:
  Optimizer failed: Ipopt: Infeasible problem detected (status 2)
[error] OPTIMIZATION FAILED...
[error] CMC::computeControls: Optimizer could not find a solution.
Unable to find a feasible solution at time = 0.`

All data including the error.log can be found here  [cmc_crash.zip](https://github.com/opensim-org/opensim-core/files/8452920/cmc_crash.zip)

It is a not too complicated  model, 13 bodies and joints, no muscles or contact forces.
Nothing special, only a constraint to close a kinematic loop.
The trajectory is generated and the trajectory.mot is created with the IK-tool. Seems OK to me.

I am completely at a loss why there is a crash directly at the first timestamp.
The positions are feasable (being created in the IK-tool) and the movement in the beginning is very slow.
What could be the cause for a crash?

UPDATE: forgot to mention: one of the coordinates (bJoint_3) is NOT controlled via bb_CMC_Tasks.xml. In fact the goal of the simulation is to find the values of this coordinate.

  Thanks, Sietse

PS. The above is done with the fast optimization. When using slow optimization it seems to work, but the model becomes unstable after a few seconds. This also when increasing the sampling rate of the trajectory from 100Hz to 200Hz.

PS2:  I asked this before for a more complicated model. But, also after studying the CMC-tool code, I don't see a reason that this type of mode could not work. Hence my asking again.

**Comment by tnamayeshi** [comment]
I have a similar issue as well.

**Comment by nickbianco** [comment]
> UPDATE: forgot to mention: one of the coordinates (bJoint_3) is NOT controlled via bb_CMC_Tasks.xml. In fact the goal of the simulation is to find the values of this coordinate.

This is probably the reason the problem fails. When running CMC, you need to have a CoordinateActuator at every joint. CMC is a tool for tracking experimental data. You cannot use it to predict the kinematics of a coordinate like you are tying to do.

**Comment by SietseAchterop** [comment]
Hello Nick,

I still don't understand it, sorry.
I do have actuators for all joints, but I exclude one from the tasks list. From the CMC documentation about the Setup_CMC file:

  `In version 3.1 and later, you can tell CMC to ignore specific actuators by adding them to the list of actuators using the property <actuators_to_exclude>.
  This is useful if you have, for example, a model component with its own controller.`

So that is what I am doing. (see PS below)
From my perspective I have my own controller for bJoint_3, it just doesn't do anything.
So nothing "illegal" as far as I can see. I have a few examples of other models that use the same idea and work fine.

To solve my problem I did simplify the model, but that led to another problem, see this  issue:
[https://github.com/opensim-org/opensim-gui/issues/1337](https://github.com/opensim-org/opensim-gui/issues/1337)
Maybe that problem is the root cause of the problem here?

If it really is not possible to use CMC, what could I do? I really want to use Simbody because of it capabilities, muscles etc.
I guess I could write my own controller for it.
In the code examples there is ControllerExample that I could use as a starting point.
But I want a controller that can cope with muscles, so I probably want to use static optimization.
So actually I want exactly what is described in "How CMC works". Unless there is a flaw in the implementation I don't see anything
that would cause a crash.

   Thanks again, Sietse
   
PS: I actually forgot the <actuator_to_exclude> in the zip file version, but I now have in bb_Setup_CMC.xml:
       <actuators_to_exclude> bJoint_3 </actuators_to_exclude>
     with the same result.
     It seems that NOT mentioning the actuator in the CMC_Tasks.xml is the way to not use that actuator in CMC.

**Comment by nickbianco** [comment]
> From my perspective I have my own controller for bJoint_3, it just doesn't do anything.

The problem is failing while using the fast target because the fast target enforces the acceleration tracking as a constraint in the optimization. Even if the target trajectory for `bJoint_3` is all zeros, you still need an actuator (with an appropriate control strategy, if not CMC-controlled) to enforce that motion (or rather, lack of motion). However, your controller isn't doing anything  so the problem can't find a solution to enforce the acceleration constraint.

You say above that the goal of the simulation is to find the value of coordinate `bJoint_3`. The only way to do this with CMC is by using the slow target, since that enforces acceleration tracking as a cost term. You can place a zero or very low weight on this coordinate tracking term in the CMC tasks file so that the problem "ignores" the tracking of this coordinate.

> When using slow optimization it seems to work, but the model becomes unstable after a few seconds. This also when increasing the sampling rate of the trajectory from 100Hz to 200Hz.

Does the trajectory look good in the initial few seconds? What do you mean by "becomes unstable"? Is the constraint violated or do the kinematics become chaotic?

**Comment by SietseAchterop** [comment]
Hello Nick,,
Sorry for my persistence, but I am not there yet.

Concerning using fast target:
Looking at the explanation in "How CMC works" I don't see the need of a control strategy for the control of bJoint_3.
CMC looks to me like a forward dynamics simulation with some extra forces/torques calculated from the feedback and static optimization.
The part of the body that contains bJoint_3 does not have this external influance and only is part of the forward dynamics part.
So it should follow the laws of physics as implemented by forward dynamics.
If I remove the actuator for bJoint_3, then it works in the same way.
In this case there is no actuator for bJoint_3 in the sum of squared controls (J) in the static optimization problem formulation,
so no enforcement of the acceleration tracking for bJoint_3.
Am I right?
I am going to study the internals of the CMC-tool in the hope to understand it better.

But for now I am focussing on the slow target method.
The simulation only becomes unstable after a few seconds. That is almost after a full rowing cycle.
The constraint (in the leg) is not violated at all. First the upper-arms a starting to rotate violantly and later also the other joints.
So only the kinematics becomes chaotic, but in such an extreme fashion that is seems that there is more wrong than a badly tuned PD controller.
Does that help?

Here the current model and data: 
[cmc_crash2.zip](https://github.com/opensim-org/opensim-core/files/8511350/cmc_crash2.zip)


cmc_results.mot is the trajectory generated by cmc.
error_slow.log is the console output during cmc processing.

  Thanks again,
    Sietse

**Comment by SietseAchterop** [comment]
Hello Nick, thanks for the detailed explanation!

I think I understand CMC now. If I want something like I thought the CMC-tool was doing, I could try implement it myself.
Maybe a bit too much for me, but maybe I will give it a try.

Concerning slow target:
>it can become unstable if the model is not strong enough to track the data, the time step is too small, etc.

I will play a bit more with weights and PD values to see whether that helps.
But what about "time step is too small"?
Do you mean the steps in the trajectory? I would think that these can't be too small.

I am still also trying to use Moco and will come back to that on the forum.

Thanks, Sietse

**Comment by nickbianco** [comment]
You're welcome!

> But what about "time step is too small"?

Sorry, I meant to say time steps are too large. But regardless, that is probably not the issue. Honestly, it's not surprising the simulation becomes unstable after a few seconds simply because there is no feedback control. Predicting kinematics even a few seconds into the future without any feedback or other control strategy will typically lead to an unstable solution.

You could try your own method, but I would strongly recommend Moco as it will be much easier to support you from my end.

**Comment by nickbianco** [comment]
CMC is all about feedback, but if you set a tracking task for a joint to zero, then the feedback for that joint is removed from the problem. The feedback controller is used to compute the accelerations that go into the tracking targets.

Once one joint becomes unstable, it's possible for the rest of the system to become unstable to the point where the other joints can track the desired kinematics. 

CMC is not typically used with contact models. Usually, we apply known ground reaction forces that correspond to measured kinematics from the same data collection, which gives CMC the best chance to find a stable solution.  If your contact model parameters/geometry don't match your kinematics well, CMC will likely have trouble finding a good solution.

**Comment by SietseAchterop** [comment]
Hello, I still am trying to get it to work.
Using Moco is probably not an option because of the complexity of my final model and the expected performance.
This together with the fact that it should be possible to get a simulation to work, with CMC (fast target)  or a custom approach.
So for now I am back with the fast-target-CMC.

I have a few examples that work fine with excluding some actuators. The optimizer then doesn't have constraints for these actuators/coordinates.
For example, see the Pusher (push_Setup_CMC.xml) example in

    https://github.com/SietseAchterop/Rowing-simulator

Two actuators are excluded and the simulation works just fine, the 2 relevant coordinates work as expected by the physics as far as I can see.
In the opensim-core source I added some code to the end of the function constraintFunc in ActuatorForceTargetFast.cpp like:
   cout << "constraints =\n" << constraints << endl;

which yields something like e.g.:
       constraints =
~[0 0 1.09139e-11]
  (also reported as the unscaled contraint violation)
So the constraints are met.

There were 5 actuators in this model, but here we see only the 3 that are not excluded. The other 2 are not interfered with by the static optimization part.
The (rest of the) simulator can just do its thing with the model and the pusher is "pushed through the water".
There also is an example pusher_muscle.py, with 2 muscles to control the knee, this also works.

So in contrast to what you told me, I now assume that it is possible to use opensim to calculate an unknown coordinate when it is excluded in CMC.
These examples show it.
But is there something different with the BootBaan model, maybe the constraint for the kinematic loop?

For completness here the updated files: [cmc_crash2.zip](https://github.com/opensim-org/opensim-core/files/8878769/cmc_crash2.zip)

Looking at my BootBaan.osim model in CMC, the constraints often become something like:

 constraints =
 ~[2748.87 -1254.79 2.29505e-05 -1.25476 7.64703e-09 -2.84605e-08 1.63323e-09 -2.66573e-08 8.96762e-10]

So 9 constraints, which fits with 10 actuators of which 1 (boatact) is excluded.
Most of the time only the first 2 or 3 constraints are violated.
They are of the seat-, hip- and knee- actuator, all having to do with the kinetic loop.

In an effort to simplify my model, I removed the upperarms.
Then hopefully I would have a model that also has a constraint to test with, unlike the Pusher model.
But then I stumble upon a problem that somehow the model cannot be manipulated at all.

This is the opensim-gui issue that I opened for it:

    https://github.com/opensim-org/opensim-gui/issues/1337

Can it be that these two problems are related or have the same root cause?

   Thanks again,
       Sietse
       
PS. In the begin there also is:
       Assembly error tolerance achieved: 1.5682257059952827e-08 required: 1e-10..Model relaxing constraints and trying again.
    After which this second try seems to succeed.
    Is this also relevant?
